test/indexer/test_indexer_controller.py

```diff
@@ -0,0 +1,160 @@
+        bundle_fqids = [
+            BundleFQID('56a338fe-7554-4b5d-96a2-7df127a7640b', '2018-03-28T151023.074974Z'),
+            BundleFQID('b2216048-7eaa-45f4-8077-5a3fb4204953', '2018-03-29T104041.822717Z')
+        ]
+        for bundle_fqid in bundle_fqids:
+            notification = self.client.synthesize_notification(self.catalog, bundle_fqid)
+            event.append(self._mock_sqs_record(action='add',
+                                               catalog=self.catalog,
+                                               notification=notification))
+            bundle = self._load_canned_bundle(bundle_fqid)
+            bundles.append(bundle)
+            # Invoke the service once to produce a set of expected entities so
+            # we don't need to hard-code them. Keep in mind that this test is
+            # not covering the service, only the controller.
+            contributions = self.index_service.transform(self.catalog, bundle, delete=False)
+            expected_entities.update(
+                (c.entity.entity_id, c.entity.entity_type)
+                for c in contributions
+            )
+
+        mock_plugin = mock.MagicMock()
+        mock_plugin.fetch_bundle.side_effect = bundles
+        with mock.patch.object(IndexController,
+                               'repository_plugin',
+                               return_value=mock_plugin):
+            # Test contribution
+            self.controller.contribute(event)
+            tallies = [
+                json.loads(m.body)
+                for m in self.queue_manager.read_messages(self.controller._tallies_queue())
+            ]
+            entities_from_tallies = {
+                (t['entity_id'], t['entity_type'])
+                for t in tallies
+            }
+            self.assertSetEqual(expected_entities, entities_from_tallies)
+            self.assertListEqual([mock.call(f) for f in bundle_fqids], mock_plugin.fetch_bundle.mock_calls)
+
+            # Test aggregation for tallies, inspect for deferred tallies
+            event = [self._mock_sqs_record(**t) for t in tallies]
+            self.controller.aggregate(event)
+            messages = self.queue_manager.read_messages(self.controller._tallies_queue())
+
+            # Check that aggregation of project entity was deferred
+            project_tally = json.loads(one(messages).body)
+            expected_tally = {
+                'catalog': 'test',
+                'entity_type': 'projects',
+                'entity_id': '93f6a42f-1790-4af4-b5d1-8c436cb6feae',
+                'num_contributions': 2
+            }
+            self.assertDictEqual(project_tally, expected_tally)
+
+            # Test aggregation of deferred project entity
+            event = [self._mock_sqs_record(**project_tally)]
+            self.controller.aggregate(event)
+            messages = self.queue_manager.read_messages(self.controller._tallies_queue())
+            self.assertEqual(0, len(messages))
```

Comment:
There's probably more you can assert about the message.

Reply:
As there are no messages returned after the deferral was processed what other assertions were you thinking about?

Should we inspect the index for the number of total contributions made?

Reply:
Nah. I was confused.